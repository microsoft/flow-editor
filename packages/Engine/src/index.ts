export * from "./components";
export * from "./contexts";
export * from "./built-in";
export * from "./utils";
export * from "./hooks";
export * from "./higherOrderComponents";
export * from "./testAPI";
export * from "./Graph.interface";
export * from "./props-api/IPropsAPI";
export * from "./Features";
export * from "./models/EdgeModel";
export * from "./models/GraphModel";
export * from "./models/NodeModel";
export * from "./models/viewport";
export * from "./common/GraphEvent.constant";
export * from "./reducers/composeReducers";
