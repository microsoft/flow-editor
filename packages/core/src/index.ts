export * from "./lib/entity";
export * from "./lib/property";
