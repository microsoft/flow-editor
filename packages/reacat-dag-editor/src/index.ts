export * from "./lib/components";
export * from "./lib/contexts";
export * from "./lib/built-in";
export * from "./lib/utils";
export * from "./lib/hooks";
export * from "./lib/testAPI";
export * from "./lib/Features";
export * from "./lib/models/EdgeModel";
export * from "./lib/models/GraphModel";
export * from "./lib/models/NodeModel";
export * from "./lib/reducers/composeReducers";
export * from "./lib/models/status";
export * from "./lib/models/port";
export * from "./lib/models/node";
export * from "./lib/models/edge";
export * from "./lib/models/canvas";
export * from "./lib/models/geometry";
export * from "./lib/models/state";
export * from "./lib/models/event";
export * from "./lib/models/config/GraphConfigBuilder";
export * from "./lib/models/config/types";
export * from "./lib/createGraphState";
