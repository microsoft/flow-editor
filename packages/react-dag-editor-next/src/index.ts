// components
export * from "./components/ReactDagEditor";
export * from "./components/Canvas";
export * from "./components/Toolbar";
export * from "./components/Palette";

// viewmodels
export * from "./viewmodels/SingleGraphViewModel";
