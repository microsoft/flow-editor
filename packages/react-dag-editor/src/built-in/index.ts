export * from "./rect";
export * from "./line";
export * from "./defaultPort";
export * from "./defaultClipboard";
export * from "./defaultStorage";
export * from "./applyDefaultPortsPosition";
export * from "./applyHorizontalGraphPortsPosition";
