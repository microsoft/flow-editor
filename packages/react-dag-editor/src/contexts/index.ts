export * from "./GraphConfigContext";
export * from "./ContextMenuConfigContext";
export * from "./ViewportContext";
export * from "./GraphStateContext";
