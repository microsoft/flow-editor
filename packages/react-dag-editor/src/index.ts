export * from "./lib/components";
export * from "./lib/contexts";
export * from "./lib/built-in";
export * from "./lib/utils";
export * from "./lib/hooks";
export * from "./lib/testAPI";
export * from "./lib/Features";
export * from "./lib/reducers/composeReducers";
export * from "./lib/models/status";
export * from "./lib/models/port";
export * from "./lib/models/node";
export * from "./lib/models/edge";
export * from "./lib/models/canvas";
export * from "./lib/models/ContentState";
export * from "./lib/models/geometry";
export * from "./lib/models/state";
export * from "./lib/models/event";
export * from "./lib/models/config/GraphConfig";
export * from "./lib/models/config/GraphConfigBuilder";
export * from "./lib/models/config/types";
export * from "./lib/createGraphState";

export * as ContentUtils from "./lib/content-utils";
export * as NodeUtils from "./lib/node-utils";
