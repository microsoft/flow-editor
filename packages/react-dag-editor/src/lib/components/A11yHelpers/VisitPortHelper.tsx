import * as React from "react";
import { GraphModel } from "../../models/GraphModel";

interface INeighborItem {
  nodeId: string;
  portId: string;
}

interface IConnectionHelperProps {
  data: GraphModel;
  neighborPorts: INeighborItem[];
  onComplete(item: INeighborItem): void;
}

export const VisitPortHelper: React.FunctionComponent<IConnectionHelperProps> = props => {
  const { neighborPorts, data } = props;
  const selectRef = React.useRef<HTMLSelectElement>(null);

  const [selectedItem, setSelectedItem] = React.useState<INeighborItem>();

  const onContainerKeyDown: React.KeyboardEventHandler = React.useCallback(
    evt => {
      if (evt.key === "Escape") {
        evt.stopPropagation();
        evt.preventDefault();

        if (selectedItem) {
          props.onComplete(selectedItem);
        }
      }
    },
    [selectedItem, props]
  );

  const onContainerBlur: React.FocusEventHandler = React.useCallback(evt => {
    //
  }, []);

  const onContainerChange: React.ChangeEventHandler<HTMLSelectElement> = React.useCallback(
    evt => {
      const value = JSON.parse(evt.target.value);

      if (value.nodeId && value.portId) {
        setSelectedItem({ nodeId: value.nodeId, portId: value.portId });
      }
    },
    [setSelectedItem]
  );

  React.useEffect(() => {
    if (selectRef.current) {
      selectRef.current.focus({ preventScroll: true });
    }
  }, []);

  return (
    <select
      onKeyDown={onContainerKeyDown}
      onBlur={onContainerBlur}
      ref={selectRef}
      onChange={onContainerChange}
    >
      {neighborPorts.map(s => {
        const isSelected =
          selectedItem &&
          selectedItem.portId === s.portId &&
          selectedItem.nodeId === s.nodeId;

        const value = JSON.stringify(s);
        const node = data.nodes.get(s.nodeId);

        if (!node) {
          return null;
        }

        const port = node.ports
          ? node.ports.filter(p => p.id === s.portId)[0]
          : null;

        if (!port) {
          return null;
        }

        const label = `${node.ariaLabel ||
          node.name ||
          node.id}: ${port.ariaLabel || port.name || port.id}`;

        return (
          <option
            key={`${s.nodeId}-${s.portId}`}
            value={value}
            aria-selected={isSelected}
            aria-label={label}
          >
            {label}
          </option>
        );
      })}
    </select>
  );
};
