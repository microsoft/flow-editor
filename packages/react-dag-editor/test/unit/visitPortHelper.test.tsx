import * as React from "react";
import { cleanup, render, fireEvent } from "@testing-library/react";
import { act } from "react-dom/test-utils";
import { VisitPortHelper } from "../../src/components/A11yHelpers/VisitPortHelper";
import { getNeighborPorts, GraphModel } from "../../src";
import { getSample4Data } from "./__data__/getSample4Data";

describe("test VisitPortHelper", () => {
  afterEach(cleanup);

  const data = GraphModel.fromJSON(getSample4Data());
  // let onCompleteSpy;

  it("the port has no neighbor port", () => {
    const neighborPorts = getNeighborPorts(data, "24a8e253", "port-2");
    const { container } = render(<VisitPortHelper neighborPorts={neighborPorts} onComplete={jest.fn()} data={data} />);
    expect(container).toMatchSnapshot();
  });

  it("the port has one neighbor port", () => {
    const neighborPorts = getNeighborPorts(data, "24a8e253", "port-1");
    const { container } = render(<VisitPortHelper neighborPorts={neighborPorts} onComplete={jest.fn()} data={data} />);
    expect(container).toMatchSnapshot();
  });

  it("the port has multi neighbor ports", () => {
    const onComplete = jest.fn();

    const neighborPorts = getNeighborPorts(data, "8b2714dd", "port-1");
    const { container } = render(<VisitPortHelper neighborPorts={neighborPorts} onComplete={onComplete} data={data} />);
    expect(container).toMatchSnapshot();

    const selectEl = container.querySelector("select")!;
    expect(selectEl?.length).toBe(3);

    act(() => {
      fireEvent.change(selectEl, {
        target: {
          value: JSON.stringify({
            nodeId: "fc9f8703",
            portId: "port-0"
          })
        }
      });
    });
    act(() => {
      fireEvent.keyDown(selectEl, {
        key: "Escape"
      });
    });

    expect(onComplete).toBeCalledWith({ nodeId: "fc9f8703", portId: "port-0" });
  });
});
