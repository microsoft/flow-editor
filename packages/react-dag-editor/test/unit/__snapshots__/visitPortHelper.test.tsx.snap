// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`test VisitPortHelper the port has multi neighbor ports 1`] = `
<div>
  <select>
    <option
      aria-label="Execute Python Script: Dataset1"
      value="{\\"nodeId\\":\\"fc9f8703\\",\\"portId\\":\\"port-0\\"}"
    >
      Execute Python Script: Dataset1
    </option>
    <option
      aria-label="Normalize Data: Dataset"
      value="{\\"nodeId\\":\\"24a8e253\\",\\"portId\\":\\"port-0\\"}"
    >
      Normalize Data: Dataset
    </option>
    <option
      aria-label="Train Model: Dataset"
      value="{\\"nodeId\\":\\"2a1e48b4\\",\\"portId\\":\\"port-1\\"}"
    >
      Train Model: Dataset
    </option>
  </select>
</div>
`;

exports[`test VisitPortHelper the port has no neighbor port 1`] = `
<div>
  <select />
</div>
`;

exports[`test VisitPortHelper the port has one neighbor port 1`] = `
<div>
  <select>
    <option
      aria-label="Train Model: Dataset"
      value="{\\"nodeId\\":\\"20268eed\\",\\"portId\\":\\"port-1\\"}"
    >
      Train Model: Dataset
    </option>
  </select>
</div>
`;
